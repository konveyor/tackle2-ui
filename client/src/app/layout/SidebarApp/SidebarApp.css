.perspective {
  padding: var(--pf-v5-c-nav__link--PaddingTop)
    var(--pf-v5-c-nav__link--PaddingRight) var(--pf-v5-global--spacer--lg)
    var(--pf-v5-c-nav__link--PaddingLeft);
  border-bottom: 1px solid var(--pf-v5-c-nav__link--before--BorderColor);
}

.perspective .pf-v5-c-select .pf-v5-c-select__toggle {
  color: var(--pf-v5-global--BackgroundColor--100);
  background-color: var(--pf-v5-global--Color--100);
}

.perspective .pf-v5-c-select .pf-v5-c-select__toggle::before {
  border-color: var(--pf-v5-global--BackgroundColor--dark-400)
    var(--pf-v5-global--BackgroundColor--dark-400)
    var(--pf-v5-c-select__toggle--before--BorderBottomColor)
    var(--pf-v5-global--BackgroundColor--dark-400);
}

.perspective .pf-v5-c-select .pf-v5-c-select__toggle:focus::before,
.perspective .pf-v5-c-select .pf-v5-c-select__toggle:hover::before {
  border-bottom-color: var(--pf-v5-c-select__toggle--before--BorderBottomColor);
}

.perspective .pf-v5-c-select .pf-v5-c-select__toggle:focus::before {
  border-bottom-width: var(
    --pf-v5-c-select__toggle--m-expanded--before--BorderBottomWidth
  );
}
