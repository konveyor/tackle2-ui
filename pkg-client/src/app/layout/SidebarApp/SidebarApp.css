.perspective {
  padding: var(--pf-c-nav__link--PaddingTop) var(--pf-c-nav__link--PaddingRight)
    var(--pf-global--spacer--lg) var(--pf-c-nav__link--PaddingLeft);
  border-bottom: 1px solid var(--pf-c-nav__link--before--BorderColor);
}

.perspective .pf-c-select .pf-c-select__toggle {
  color: var(--pf-global--BackgroundColor--100);
  background-color: var(--pf-global--Color--100);
}

.perspective .pf-c-select .pf-c-select__toggle::before {
  border-color: var(--pf-global--BackgroundColor--dark-400)
    var(--pf-global--BackgroundColor--dark-400)
    var(--pf-c-select__toggle--before--BorderBottomColor)
    var(--pf-global--BackgroundColor--dark-400);
}

.perspective .pf-c-select .pf-c-select__toggle:focus::before,
.perspective .pf-c-select .pf-c-select__toggle:hover::before {
  border-bottom-color: var(--pf-c-select__toggle--before--BorderBottomColor);
}

.perspective .pf-c-select .pf-c-select__toggle:focus::before {
  border-bottom-width: var(
    --pf-c-select__toggle--m-expanded--before--BorderBottomWidth
  );
}
