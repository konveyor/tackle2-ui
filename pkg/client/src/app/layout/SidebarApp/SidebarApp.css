.perspective.pf-c-select {
  --pf-global--Color--100: #fff;
  --pf-global--BackgroundColor--100: #000;
}
