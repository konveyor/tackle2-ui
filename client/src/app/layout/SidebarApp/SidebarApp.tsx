import React, { useEffect } from "react";
import { NavLink, useHistory, useLocation } from "react-router-dom";
import { useTranslation } from "react-i18next";
import {
  Nav,
  NavItem,
  PageSidebar,
  NavList,
  Select,
  SelectOption,
  SelectVariant,
  NavExpandable,
} from "@patternfly/react-core";

import { Paths } from "@app/Paths";
import { LayoutTheme } from "../LayoutUtils";
import { checkAccess } from "@app/common/rbac-utils";
import keycloak from "@app/keycloak";

import { useLocalStorage } from "@migtools/lib-ui";
import { LocalStorageKey } from "@app/Constants";
import { FEATURES_ENABLED } from "@app/FeatureFlags";
import { OptionWithValue, SimpleSelect } from "@app/shared/components";
import { toOptionLike } from "@app/utils/model-utils";
import "./SidebarApp.css";

export const SidebarApp: React.FC = () => {
  const token = keycloak.tokenParsed || undefined;
  const userRoles = token?.realm_access?.roles || [],
    adminAccess = checkAccess(userRoles, ["tackle-admin"]);

  const { t } = useTranslation();
  const { search } = useLocation();
  const history = useHistory();
  enum PersonaKey {
    ADMINISTRATION = "Administration",
    MIGRATION = "Migration",
  }

  const options = [
    <SelectOption
      key="dev"
      component="button"
      value={PersonaKey.MIGRATION}
      isPlaceholder
    >
      {PersonaKey.MIGRATION}
    </SelectOption>,
    ...(adminAccess
      ? [
          <SelectOption
            key="admin"
            component="button"
            value={PersonaKey.ADMINISTRATION}
          >
            {PersonaKey.ADMINISTRATION}
          </SelectOption>,
        ]
      : []),
  ];

  const personaOptions: OptionWithValue<string>[] = [
    {
      value: PersonaKey.MIGRATION,
      toString: () => PersonaKey.MIGRATION,
    },
    {
      value: PersonaKey.ADMINISTRATION,
      toString: () => PersonaKey.ADMINISTRATION,
    },
  ];

  const [selectedPersona, setSelectedPersona] =
    useLocalStorage<PersonaKey | null>(LocalStorageKey.selectedPersona, null);

  useEffect(() => {
    if (!selectedPersona) {
      history.push("/applications");
      setSelectedPersona(PersonaKey.MIGRATION);
    }
  }, []);

  const Navigation = (
    <Nav id="nav-primary" aria-label="Nav" theme={LayoutTheme}>
      <div className="perspective">
        <SimpleSelect
          toggleId="sidebar-perspective-toggle"
          variant={SelectVariant.single}
          aria-label="Select user perspective"
          id="sidebar-perspective"
          value={
            selectedPersona
              ? toOptionLike(selectedPersona, personaOptions)
              : undefined
          }
          options={personaOptions}
          onChange={(selection) => {
            const selectionValue = selection as OptionWithValue<PersonaKey>;
            setSelectedPersona(selectionValue.value);
            if (selectionValue.value === PersonaKey.ADMINISTRATION) {
              history.push(Paths.general);
            } else {
              history.push(Paths.applications);
            }
          }}
        />
      </div>
      {selectedPersona === PersonaKey.MIGRATION ? (
        <NavList title="Global">
          <NavItem>
            <NavLink
              to={Paths.applications + search}
              activeClassName="pf-m-current"
            >
              {t("sidebar.applicationInventory")}
            </NavLink>
          </NavItem>
          <NavItem>
            <NavLink to={Paths.reports + search} activeClassName="pf-m-current">
              {t("sidebar.reports")}
            </NavLink>
          </NavItem>
          <NavItem>
            <NavLink to={Paths.controls} activeClassName="pf-m-current">
              {t("sidebar.controls")}
            </NavLink>
          </NavItem>
          {FEATURES_ENABLED.migrationWaves ? (
            <NavItem>
              <NavLink to={Paths.waves} activeClassName="pf-m-current">
                {t("sidebar.waves")}
              </NavLink>
            </NavItem>
          ) : null}
          {FEATURES_ENABLED.dynamicReports ? (
            <>
              <NavItem>
                <NavLink to={Paths.issues} activeClassName="pf-m-current">
                  {t("sidebar.issues")}
                </NavLink>
              </NavItem>
              <NavItem>
                <NavLink to={Paths.dependencies} activeClassName="pf-m-current">
                  {t("sidebar.dependencies")}
                </NavLink>
              </NavItem>
            </>
          ) : null}
        </NavList>
      ) : (
        <NavList title="Admin">
          <NavItem>
            <NavLink to={Paths.general} activeClassName="pf-m-current">
              {t("terms.general")}
            </NavLink>
          </NavItem>
          <NavItem>
            <NavLink to={Paths.identities} activeClassName="pf-m-current">
              {t("terms.credentials")}
            </NavLink>
          </NavItem>
          <NavExpandable
            title="Repositories"
            srText="SR Link"
            groupId="admin-repos"
            isExpanded
          >
            <NavItem>
              <NavLink
                to={Paths.repositoriesGit}
                activeClassName="pf-m-current"
              >
                Git
              </NavLink>
            </NavItem>
            <NavItem>
              <NavLink
                to={Paths.repositoriesSvn}
                activeClassName="pf-m-current"
              >
                Subversion
              </NavLink>
            </NavItem>
            <NavItem>
              <NavLink
                to={Paths.repositoriesMvn}
                activeClassName="pf-m-current"
              >
                Maven
              </NavLink>
            </NavItem>
          </NavExpandable>
          <NavItem>
            <NavLink to={Paths.proxies} activeClassName="pf-m-current">
              Proxy
            </NavLink>
          </NavItem>
          <NavItem>
            <NavLink to={Paths.migrationTargets} activeClassName="pf-m-current">
              Custom migration targets
            </NavLink>
          </NavItem>
          {FEATURES_ENABLED.migrationWaves ? (
            <NavExpandable
              title="Issue management"
              srText="SR Link"
              groupId="admin-issue-management"
              isExpanded
            >
              <NavItem>
                <NavLink to={Paths.jira} activeClassName="pf-m-current">
                  Jira
                </NavLink>
              </NavItem>
            </NavExpandable>
          ) : null}
        </NavList>
      )}
    </Nav>
  );

  return <PageSidebar nav={Navigation} theme={LayoutTheme} />;
};
