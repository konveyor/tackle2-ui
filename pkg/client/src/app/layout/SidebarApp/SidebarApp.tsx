import React from "react";
import { NavLink, useHistory, useLocation } from "react-router-dom";
import { useTranslation } from "react-i18next";
import {
  Nav,
  NavItem,
  PageSidebar,
  NavList,
  Select,
  SelectOption,
  SelectVariant,
  NavExpandable,
} from "@patternfly/react-core";
import AdminIcon from "@patternfly/react-icons/dist/esm/icons/cogs-icon";
import DevIcon from "@patternfly/react-icons/dist/esm/icons/code-icon";

import { Paths } from "@app/Paths";
import { LayoutTheme } from "../LayoutUtils";
import "./SidebarApp.css";

export const SidebarApp: React.FC = () => {
  const { t } = useTranslation();
  const { search } = useLocation();
  const history = useHistory();

  const onAdminClick = () => {
    console.log("Admin Selected");
  };

  const onDevClick = () => {
    console.log("Dev Selected");
  };

  const options = [
    <SelectOption
      key="dev"
      component="button"
      onClick={onDevClick}
      value="Developer"
      isPlaceholder
    />,
    <SelectOption
      key="admin"
      component="button"
      onClick={onAdminClick}
      value="Administrator"
    />,
  ];

  const [isOpen, setIsOpen] = React.useState(false);
  const [selected, setSelected] = React.useState("Developer");
  const [isDevIcon, setDevIcon] = React.useState(true);

  const onTogglePerspective = () => {
    setIsOpen(!isOpen);
  };

  const onSelectPerspective = (_, selection) => {
    setSelected(selection);
    setIsOpen(!isOpen);
    if (selection === "Administrator") {
      setDevIcon(false);
      history.push(Paths.identities);
    } else {
      setDevIcon(true);
      history.push(Paths.applications);
    }
  };

  const Navigation2 = (
    <Nav id="nav-primary-simple" theme="dark">
      <NavList id="nav-list-simple">
        {/* {routes.map(
          (route, idx) =>
            route.label && (!route.routes ? renderNavItem(route, idx) : renderNavGroup(route, idx))
        )} */}
      </NavList>
    </Nav>
  );

  const Navigation = (
    <>
      <Select
        toggleIcon={isDevIcon ? <DevIcon /> : <AdminIcon />}
        className="perspective"
        variant={SelectVariant.single}
        aria-label="Select user perspective"
        selections={selected}
        isOpen={isOpen}
        onSelect={onSelectPerspective}
        onToggle={onTogglePerspective}
      >
        {options}
      </Select>
      {selected === "Developer" ? (
        <Nav id="nav-primary" aria-label="Nav" theme={LayoutTheme}>
          <NavList title="Global">
            <NavItem>
              <NavLink
                to={Paths.applications + search}
                activeClassName="pf-m-current"
              >
                {t("sidebar.applicationInventory")}
              </NavLink>
            </NavItem>
            <NavItem>
              <NavLink
                to={Paths.reports + search}
                activeClassName="pf-m-current"
              >
                {t("sidebar.reports")}
              </NavLink>
            </NavItem>
            <NavItem>
              <NavLink to={Paths.controls} activeClassName="pf-m-current">
                {t("sidebar.controls")}
              </NavLink>
            </NavItem>
          </NavList>
        </Nav>
      ) : (
        <Nav id="nav-admin" aria-label="NavAdmin" theme={LayoutTheme}>
          <NavList title="Admin">
            <NavItem>
              <NavLink to={Paths.identities} activeClassName="pf-m-current">
                {t("terms.credentials")}
              </NavLink>
            </NavItem>
            <NavExpandable
              title="Repositories"
              srText="SR Link"
              groupId="admin-repos"
              isExpanded
            >
              <NavItem>
                <NavLink
                  to={Paths.repositoriesGit}
                  activeClassName="pf-m-current"
                >
                  Git
                </NavLink>
              </NavItem>
              <NavItem>
                <NavLink
                  to={Paths.repositoriesSvn}
                  activeClassName="pf-m-current"
                >
                  Subversion
                </NavLink>
              </NavItem>
              <NavItem>
                <NavLink
                  to={Paths.repositoriesMvn}
                  activeClassName="pf-m-current"
                >
                  Maven
                </NavLink>
              </NavItem>
            </NavExpandable>
            <NavItem>
              <NavLink to={Paths.proxies} activeClassName="pf-m-current">
                Proxy
              </NavLink>
            </NavItem>
          </NavList>
        </Nav>
      )}
    </>
  );

  return <PageSidebar nav={Navigation} theme={LayoutTheme} />;
};
