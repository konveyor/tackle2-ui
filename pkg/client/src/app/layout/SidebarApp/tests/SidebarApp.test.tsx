import React from "react";
import { BrowserRouter as Router } from "react-router-dom";
import { shallow } from "enzyme";
import { SidebarApp } from "../SidebarApp";

it("Renders without crashing", () => {
  jest.mock("react-i18next", () => ({
    useTranslation: () => {
      return {
        t: (str: any) => str,
        i18n: {
          changeLanguage: () => new Promise(() => {}),
        },
      };
    },
  }));

  const wrapper = shallow(
    <Router>
      <SidebarApp />
    </Router>
  );
  expect(wrapper).toMatchSnapshot();
});
