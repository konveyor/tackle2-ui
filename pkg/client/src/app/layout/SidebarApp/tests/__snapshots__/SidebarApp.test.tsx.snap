// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders without crashing 1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <SidebarApp />
</Router>
`;
